Zone5/Views/ThirdPartyConnectionsView.swift

```diff
@@ -9,52 +9,76 @@
-		return post(Endpoints.setThirdPartyConnection, parameters: queryParams(type), body: connection, with: completion)
-	}
-
-	@discardableResult
-	public func hasThirdPartyToken(type: UserConnectionType, completion: @escaping Zone5.ResultHandler<ThirdPartyTokenResponse>) -> PendingRequest? {
-		return get(Endpoints.hasThirdPartyConnection, parameters: queryParams(type), with: completion)
-	}
-
-	@discardableResult
-	public func removeThirdPartyToken(type: UserConnectionType, completion: @escaping Zone5.ResultHandler<ThirdPartyResponse>) -> PendingRequest? {
-		return post(Endpoints.removeThirdPartyConnection, parameters: queryParams(type), body: nil, with: completion)
-	}
+    private enum Endpoints: String, InternalRequestEndpoint {
+        case initializePairing = "/rest/users/connections/pair/{connectionType}"
+        case confirmConnection = "/rest/files/{connectionType}/confirm"
+        case userConnections = "/rest/users/connections"
+        case removeThirdPartyConnection = "/rest/users/connections/rem/{connectionType}"
+        case registerDeviceWithThirdParty = "/rest/users/scheduled/activities/api/v1/push_registrations"
+        case deregisterDeviceWithThirdParty = "/rest/users/scheduled/activities/api/v1/push_registrations/{token}"
+    }
+
+    private let serviceKey: String = "service_name"
+    internal func queryParams(_ serviceType: UserConnectionType) -> URLEncodedBody {
+        let queryParams: URLEncodedBody = [ serviceKey : "\(serviceType)" ]
+        return queryParams
+    }
+
+    /// Register a push token for a device with a 3rd party
+    ///- Parameters:
+    /// - PushRegistration (token, platform, deviceId): Push registration for a device with a third party
+    @discardableResult
+    public func registerDeviceWithThirdParty(registration: PushRegistration, completion: @escaping Zone5.ResultHandler<PushRegistrationResponse>) -> PendingRequest? {
+        return post(Endpoints.registerDeviceWithThirdParty, body: registration, with: completion)
+    }
+
+    /// Deregister a push token for a device with a 3rd party
+    /// - Parameters:
+    /// - token: 3rd party push token to deregister
+    @discardableResult
+    public func deregisterDeviceWithThirdParty(token: String, completion: @escaping Zone5.ResultHandler<Zone5.VoidReply>) -> PendingRequest? {
+        let endpoint = Endpoints.deregisterDeviceWithThirdParty.replacingTokens(["token": token])
+        return delete(endpoint, with: completion)
+    }
+
+    /// Initialize a connection for the current user for the given 3rd party type
+    /// - Parameters
+    @discardableResult
+    public func initializeThirdPartyConnection(type: UserConnectionType, completion: @escaping Zone5.ResultHandler<ThirdPartyInitializeResponse>) -> PendingRequest? {
+        let endpoint = Endpoints.initializePairing.replacingTokens(["connectionType": type.connectionName])
+        return post(endpoint, body: EmptyBody(), with: completion)
+    }
```

Comment:
Do you need to send an empty body? Does it work if you just call `post(endpoint, body: nil, with: completion)`? body is an optional. You'll note that the old removeThirdPartyToken code did this. If you can just pass nil then the new EmptyBody struct is unnecessary.

Reply:
Yes, this is needed. I tested with a nil and it returns a failure (decode error). James told me to send an empty body like this: "{}"

Reply:
ok